driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -0,0 +1,524 @@
+        }
+    }
+
+    @Override
+    public void next(final SingleResultCallback<List<T>> callback) {
+        if (isClosed()) {
+            callback.onResult(null, new MongoException(MESSAGE_IF_CLOSED_AS_CURSOR));
+            return;
+        } else if (!resourceManager.tryStartOperation()) {
+            callback.onResult(null, new MongoException(MESSAGE_IF_CONCURRENT_OPERATION));
+            return;
+        }
+
+        ServerCursor localServerCursor = resourceManager.serverCursor();
+        boolean cursorClosed = localServerCursor == null;
+        List<T> batchResults = emptyList();
+        if (!processedInitial.getAndSet(true) && !firstBatchEmpty) {
+            batchResults = commandCursorResult.get().getResults();
+        }
+
+        if (cursorClosed || !batchResults.isEmpty()) {
+            resourceManager.endOperation();
+            if (cursorClosed) {
+                close();
+            }
+            callback.onResult(batchResults, null);
+        } else {
+            getMore(localServerCursor, callback);
+        }
+    }
+
+    @Override
+    public void setBatchSize(final int batchSize) {
+        this.batchSize = batchSize;
+    }
+
+    @Override
+    public int getBatchSize() {
+        return batchSize;
+    }
+
+    @Override
+    public boolean isClosed() {
+        return !resourceManager.operable();
+    }
+
+    @Override
+    public void close() {
+        resourceManager.close();
+    }
+
+    @Override
+    public BsonDocument getPostBatchResumeToken() {
+        return commandCursorResult.get().getPostBatchResumeToken();
+    }
+
+    @Override
+    public BsonTimestamp getOperationTime() {
+        return commandCursorResult.get().getOperationTime();
+    }
+
+    @Override
+    public boolean isFirstBatchEmpty() {
+        return firstBatchEmpty;
+    }
+
+    @Override
+    public int getMaxWireVersion() {
+        return maxWireVersion;
+    }
+
+    @Nullable
+    @VisibleForTesting(otherwise = VisibleForTesting.AccessModifier.PRIVATE)
+    ServerCursor getServerCursor() {
+        return resourceManager.serverCursor();
+    }
+
+    private void getMore(final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
+        resourceManager.executeWithConnection(callback, connection -> getMore(connection, cursor, callback));
+    }
+
+    private void getMore(final AsyncConnection connection, final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
+        connection.commandAsync(namespace.getDatabaseName(),
+                getMoreCommandDocument(cursor.getId(), connection.getDescription(), namespace,
+                        limit, batchSize, count.get(), maxTimeMS, comment),
+                NO_OP_FIELD_NAME_VALIDATOR, ReadPreference.primary(),
+                CommandResultDocumentCodec.create(decoder, NEXT_BATCH), assertNotNull(resourceManager.connectionSource),
+                (commandResult, t) -> {
+                    if (t != null) {
```

Comment:
I used multiple returns to make the code more readable and mimic the sync returns, even though we are using callbacks.